File: Backpack/TextView/Classes/BPKTextView.m
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#import "BPKTextView.h"

#import <Backpack/Color.h>
#import <Backpack/Common.h>

NS_ASSUME_NONNULL_BEGIN
@interface BPKTextView ()
- (void)setupWithStyle:(BPKFontStyle)style;
@end

@implementation BPKTextView

- (instancetype)initWithFontStyle:(BPKFontStyle)style {
    BPKAssertMainThread();
    self = [super initWithFrame:CGRectZero textContainer:nil];

    if (self) {
        [self setupWithStyle:style];
    }

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
    BPKAssertMainThread();
    self = [super initWithCoder:aDecoder];

    if (self) {
        [self setupWithStyle:BPKFontStyleTextBodyDefault];
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame textContainer:(nullable NSTextContainer *)textContainer {
    BPKAssertMainThread();
    self = [super initWithFrame:frame textContainer:textContainer];

    if (self) {
        [self setupWithStyle:BPKFontStyleTextBodyDefault];
    }

    return self;
}

- (void)setText:(NSString *_Nullable)text {
    BPKAssertMainThread();
    if (text == nil) {
        self.attributedText = nil;
        return;
    }

    NSAttributedString *attributedString = nil;
    if (self.textColor) {
        attributedString = [BPKFont attributedStringWithFontStyle:self.fontStyle content:text textColor:self.textColor];
    } else {
        attributedString = [BPKFont attributedStringWithFontStyle:self.fontStyle content:text];
    }
    self.attributedText = attributedString;
}

- (void)setTextColor:(UIColor *_Nullable)textColor {
    BPKAssertMainThread();
    [super setTextColor:textColor];

    [self updateStyle];
}

- (void)setFontStyle:(BPKFontStyle)fontStyle {
    BPKAssertMainThread();
    _fontStyle = fontStyle;

    [self updateStyle];
}

#pragma mark - Private

- (void)setupWithStyle:(BPKFontStyle)style {
    self.fontStyle = style;
    self.textColor = BPKColor.textPrimaryColor;

    [self updateStyle];
}

- (void)updateStyle {
    NSDictionary *colorAttributes = self.textColor ? @{NSForegroundColorAttributeName: self.textColor} : @{};
    self.typingAttributes = [BPKFont attributesForFontStyle:self.fontStyle withCustomAttributes:colorAttributes];

    [self setText:self.attributedText.string];
}

- (void)traitCollectionDidChange:(UITraitCollection * __nullable)previousTraitCollection {
    [super traitCollectionDidChange:previousTraitCollection];
    
    [self updateStyle];
}

@end
NS_ASSUME_NONNULL_END

File: Backpack/TextView/Classes/BPKTextView.swift
/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import UIKit

/// `BPKTextView` is a subclass of `UITextView` which uses the Skyscanner style.
@available(*, deprecated, message: "BPKTextView is deprecated. Use the SwiftUI BPKTextArea instead.")
@objcMembers
@IBDesignable
public final class BPKTextView: UITextView {

    /// The font style used for the TextView.
    ///
    /// See `BPKFontStyle` for the integer values to use when setting from Interface Builder.
    public var fontStyle: BPKFontStyle = .textBodyDefault {
        didSet {
            guard fontStyle != oldValue else { return }
            applyStyle()
        }
    }

    // MARK: - Initializers

    /// Create a `BPKTextView` with a specific BPKFont style.
    ///
    /// - Parameter fontStyle: Font style to be used by the TextView.
    public init(fontStyle: BPKFontStyle) {
        super.init(frame: .zero, textContainer: nil)
        commonInit(fontStyle: fontStyle)
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit(fontStyle: .textBodyDefault)
    }

    public override init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        commonInit(fontStyle: .textBodyDefault)
    }

    // MARK: - Overrides

    public override var text: String! {
        didSet {
            applyStyle()
        }
    }

    public override var textColor: UIColor? {
        didSet {
            guard textColor != oldValue else { return }
            applyStyle()
        }
    }

    // MARK: - Private

    private func commonInit(fontStyle: BPKFontStyle) {
        self.fontStyle = fontStyle
        super.textColor = BPKColor.textPrimaryColor
        applyStyle()
    }

    private func applyStyle() {
        let colorAttributes: [NSAttributedString.Key: Any] = textColor.map {
            [.foregroundColor: $0]
        } ?? [:]

        typingAttributes = BPKFont.makeAttributes(
            fontStyle: fontStyle,
            customAttributes: colorAttributes
        )

        guard let currentText = super.text, !currentText.isEmpty else { return }
        guard let color = textColor else { return }

        super.attributedText = BPKFont.makeAttributedString(
            fontStyle: fontStyle,
            content: currentText,
            textColor: color
        )
    }
}
